node/src/accountant/mod.rs

```diff
@@ -1236,38 +1429,150 @@ impl From<&PendingPayableFingerprint> for PendingPayableId {
     }
 
-    fn is_innocent_balance(&self, balance: i64, limit: i64) -> bool {
+    fn is_innocent_balance(&self, balance: u128, limit: u128) -> bool {
         balance <= limit
     }
 
-    fn calculate_payout_threshold(&self, payment_thresholds: PaymentThresholds, x: u64) -> f64 {
-        let m = -((payment_thresholds.debt_threshold_gwei as f64
-            - payment_thresholds.permanent_debt_allowed_gwei as f64)
-            / (payment_thresholds.threshold_interval_sec as f64
-                - payment_thresholds.maturity_threshold_sec as f64));
-        let b = payment_thresholds.debt_threshold_gwei as f64
-            - m * payment_thresholds.maturity_threshold_sec as f64;
-        m * x as f64 + b
+    fn calculate_payout_threshold(&self, payment_thresholds: &PaymentThresholds, x: u64) -> u64 {
+        ThresholdUtils::calculate_sloped_threshold_by_time(payment_thresholds, x) as u64
     }
     as_any_impl!();
 }
 
+pub struct ThresholdUtils {}
+
+impl ThresholdUtils {
+    pub fn slope(payment_thresholds: &PaymentThresholds, for_wei_computation: bool) -> f64 {
+        if payment_thresholds.debt_threshold_gwei <= payment_thresholds.permanent_debt_allowed_gwei
+        {
+            panic!(
+                "Debt threshold ({}) cannot be smaller than permanent debt allowed ({})",
+                payment_thresholds.debt_threshold_gwei,
+                payment_thresholds.permanent_debt_allowed_gwei
+            )
+        }
+        let slope = -(payment_thresholds.debt_threshold_gwei as f64
+            - payment_thresholds.permanent_debt_allowed_gwei as f64)
+            / Self::convert(payment_thresholds.threshold_interval_sec) as f64;
+        if for_wei_computation {
+            slope * 10_u64.pow(9) as f64
+        } else {
+            slope
+        }
+    }
+
+    fn convert(num: u64) -> i64 {
+        checked_conversion::<u64, i64>(num)
+    }
+
+    fn compute_theoretical_interception_with_y_axis(
+        m: f64,
+        debt_threshold: i64,
+        maturity_threshold_sec: i64,
+    ) -> f64 {
+        let m = m.abs();
+        let x_interception = (debt_threshold as f64 / m) + maturity_threshold_sec as f64;
+        m * x_interception
+    }
+
+    pub fn calculate_sloped_threshold_by_time(
+        payment_thresholds: &PaymentThresholds,
+        time: u64,
+    ) -> f64 {
+        let m = ThresholdUtils::slope(payment_thresholds, false);
+        let b = ThresholdUtils::compute_theoretical_interception_with_y_axis(
+            m,
+            ThresholdUtils::convert(payment_thresholds.debt_threshold_gwei),
+            ThresholdUtils::convert(payment_thresholds.maturity_threshold_sec),
+        );
+        let y = m * time as f64 + b;
+        let f_debt_threshold_gwei = payment_thresholds.debt_threshold_gwei as f64;
+        let f_permanent_debt_allowed_gwei = payment_thresholds.permanent_debt_allowed_gwei as f64;
+        if y >= f_debt_threshold_gwei {
+            return f_debt_threshold_gwei;
```

Comment:
This is wrong. I think this shouldn't ever happen, because x values that produce y values in this range should not come into this function. This should be modified either to be right (either no limit or f64::MAX) or to panic with a message about bad code.